File: articles/probability-for-systems-modeling/r/monte_carlo_risk.R
# Probability for Systems Modeling:
# Monte Carlo risk simulation in R.
# Educational example only.

library(tidyverse)

set.seed(42)

parameters <- read_csv("../data/monte_carlo_risk_parameters.csv", show_col_types = FALSE)

simulate_system_loss <- function(exposure_min, exposure_max, vulnerability_alpha, vulnerability_beta, shock_meanlog, shock_sdlog, iterations) {
  tibble(
    exposure = runif(iterations, min = exposure_min, max = exposure_max),
    vulnerability = rbeta(iterations, shape1 = vulnerability_alpha, shape2 = vulnerability_beta),
    shock_intensity = rlnorm(iterations, meanlog = shock_meanlog, sdlog = shock_sdlog)
  ) |>
    mutate(
      system_loss = exposure * vulnerability * shock_intensity
    )
}

simulation_results <- parameters |>
  mutate(
    simulation = pmap(
      list(exposure_min, exposure_max, vulnerability_alpha, vulnerability_beta, shock_meanlog, shock_sdlog, iterations),
      simulate_system_loss
    )
  ) |>
  select(scenario_id, simulation) |>
  unnest(simulation)

summary_results <- simulation_results |>
  group_by(scenario_id) |>
  summarise(
    expected_loss = mean(system_loss),
    median_loss = median(system_loss),
    loss_sd = sd(system_loss),
    p90_loss = quantile(system_loss, 0.90),
    p95_loss = quantile(system_loss, 0.95),
    p99_loss = quantile(system_loss, 0.99),
    probability_loss_gt_0_50 = mean(system_loss > 0.50),
    .groups = "drop"
  )

dir.create("../outputs", showWarnings = FALSE, recursive = TRUE)

write_csv(simulation_results, "../outputs/r_monte_carlo_risk_results.csv")
write_csv(summary_results, "../outputs/r_monte_carlo_risk_summary.csv")

print(summary_results)

File: articles/probability-for-systems-modeling/python/monte_carlo_risk.py
"""
Probability for Systems Modeling:
Monte Carlo risk simulation.

Educational example only.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def simulate_system_loss(
    exposure_min: float,
    exposure_max: float,
    vulnerability_alpha: float,
    vulnerability_beta: float,
    shock_meanlog: float,
    shock_sdlog: float,
    iterations: int,
    seed: int
) -> pd.DataFrame:
    """Simulate uncertain system loss from exposure, vulnerability, and shock intensity."""
    rng = np.random.default_rng(seed)

    exposure = rng.uniform(exposure_min, exposure_max, iterations)
    vulnerability = rng.beta(vulnerability_alpha, vulnerability_beta, iterations)
    shock_intensity = rng.lognormal(mean=shock_meanlog, sigma=shock_sdlog, size=iterations)

    system_loss = exposure * vulnerability * shock_intensity

    return pd.DataFrame({
        "exposure": exposure,
        "vulnerability": vulnerability,
        "shock_intensity": shock_intensity,
        "system_loss": system_loss
    })


def summarize_loss(simulation: pd.DataFrame) -> dict[str, float]:
    """Summarize simulated loss distribution."""
    return {
        "expected_loss": float(simulation["system_loss"].mean()),
        "median_loss": float(simulation["system_loss"].median()),
        "loss_sd": float(simulation["system_loss"].std()),
        "p90_loss": float(simulation["system_loss"].quantile(0.90)),
        "p95_loss": float(simulation["system_loss"].quantile(0.95)),
        "p99_loss": float(simulation["system_loss"].quantile(0.99)),
        "probability_loss_gt_0_50": float((simulation["system_loss"] > 0.50).mean())
    }


def main() -> None:
    parameters = pd.read_csv("../data/monte_carlo_risk_parameters.csv")

    all_results = []
    summary_rows = []

    for i, row in parameters.iterrows():
        simulation = simulate_system_loss(
            exposure_min=float(row["exposure_min"]),
            exposure_max=float(row["exposure_max"]),
            vulnerability_alpha=float(row["vulnerability_alpha"]),
            vulnerability_beta=float(row["vulnerability_beta"]),
            shock_meanlog=float(row["shock_meanlog"]),
            shock_sdlog=float(row["shock_sdlog"]),
            iterations=int(row["iterations"]),
            seed=42 + i
        )

        simulation["scenario_id"] = row["scenario_id"]
        all_results.append(simulation)

        summary = summarize_loss(simulation)
        summary["scenario_id"] = row["scenario_id"]
        summary_rows.append(summary)

    results = pd.concat(all_results, ignore_index=True)
    summary = pd.DataFrame(summary_rows)

    print(summary)

    results.to_csv("../outputs/python_monte_carlo_risk_results.csv", index=False)
    summary.to_csv("../outputs/python_monte_carlo_risk_summary.csv", index=False)


if __name__ == "__main__":
    main()
